"""
Script de Teste para os Servi√ßos gRPC
Execute ap√≥s iniciar todos os microsservi√ßos
"""
import grpc
import sys
import os

# Adicionar diret√≥rio generated ao path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'generated'))

import gateway_pb2
import gateway_pb2_grpc
import leilao_pb2

def test_gateway():
    """Testar conex√£o com o Gateway"""
    print("\n=== Testando Gateway gRPC ===\n")
    
    try:
        # Conectar ao Gateway
        channel = grpc.insecure_channel('localhost:50054')
        stub = gateway_pb2_grpc.GatewayServiceStub(channel)
        
        # 1. Listar leil√µes
        print("1. Listando leil√µes...")
        response = stub.ListarLeiloes(leilao_pb2.ListarLeiloesRequest())
        print(f"   ‚úì Encontrados {len(response.leiloes)} leil√µes")
        for leilao in response.leiloes:
            print(f"   - Leil√£o {leilao.id}: {leilao.nome} - R$ {leilao.valor_inicial}")
        
        # 2. Registrar interesse
        print("\n2. Registrando interesse...")
        cliente_id = "test_user_123"
        leilao_id = 1
        response = stub.RegistrarInteresse(leilao_pb2.RegistrarInteresseRequest(
            leilao_id=leilao_id,
            cliente_id=cliente_id
        ))
        if response.success:
            print(f"   ‚úì Interesse registrado: {response.message}")
        else:
            print(f"   ‚úó Erro: {response.message}")
        
        # 3. Testar stream de notifica√ß√µes (apenas 5 segundos)
        print("\n3. Conectando ao stream de notifica√ß√µes (5 segundos)...")
        print("   (Aguardando notifica√ß√µes...)")
        
        import threading
        import time
        
        def listen_stream():
            try:
                for notificacao in stub.StreamNotificacoesUnificadas(
                    gateway_pb2.StreamNotificacoesUnificadasRequest(cliente_id=cliente_id)
                ):
                    print(f"   üì© Notifica√ß√£o: {notificacao.tipo} - Leil√£o {notificacao.leilao_id}")
            except Exception as e:
                print(f"   Stream finalizado: {e}")
        
        stream_thread = threading.Thread(target=listen_stream, daemon=True)
        stream_thread.start()
        time.sleep(5)
        
        print("   ‚úì Stream conectado com sucesso")
        
        # 4. Cancelar interesse
        print("\n4. Cancelando interesse...")
        response = stub.CancelarInteresse(leilao_pb2.CancelarInteresseRequest(
            leilao_id=leilao_id,
            cliente_id=cliente_id
        ))
        if response.success:
            print(f"   ‚úì Interesse cancelado: {response.message}")
        
        print("\n‚úÖ Todos os testes passaram!")
        print("\nüí° Dica: Para testar lances e pagamentos, use o frontend ou grpcurl")
        
        channel.close()
        
    except grpc.RpcError as e:
        print(f"\n‚ùå Erro gRPC: {e.code()} - {e.details()}")
        print("\n‚ö†Ô∏è  Certifique-se de que todos os microsservi√ßos est√£o rodando:")
        print("   - python services/ms_leilao.py")
        print("   - python services/ms_lance.py")
        print("   - python services/ms_pagamento.py")
        print("   - python app.py")
    except Exception as e:
        print(f"\n‚ùå Erro: {e}")
        print("\n‚ö†Ô∏è  Voc√™ executou a gera√ß√£o dos arquivos proto?")
        print("   python -m grpc_tools.protoc -I./protos --python_out=./generated --grpc_python_out=./generated ./protos/*.proto")


if __name__ == "__main__":
    test_gateway()
